### ISSUE TITLE
queries produce internal filenames and duplicate entries

### ISSUE THREAD

**Issue by BrainDamage** [StrumentiResistenti/Tagsistant#28]
this bug shows up with git head 2e97ac16ee284404b810332ffd2eadc5346fd7f7 but not release of 0.8.1a

this one is going to be hard to debug because the effect doesn't have an apparent trigger:
for only some files,  instead of their normal name the internal state is leaked during a query and results are duplicated, eg:
```
$ ls store/mp3/@
1___Ailsean - Subaqueous Angels of Chaos (FF I).mp3
1___Ailsean - Subaqueous Angels of Chaos (FF I).mp3
AmIEvil - Red XIII Redux (FF VII).mp3
```
as if i'd be randomly reading from the archive dir
even more curiously is that the result from the ALL metata report the correct name and only once:
```
$ ls store/ALL/@ | grep "Ailsean - Subaqueous Angels of Chaos (FF I).mp3"
Ailsean - Subaqueous Angels of Chaos (FF I).mp3
```
however, the positive news is that even if there's no apparent relation or trigger, the effect seems to be deterministic, it's always the same files producing the effect, and they do it for any valid query except for the ALL one

mounting the same dir with the stable version of tagsistant gives the correct output, making it a regression since the release

I've attempted a bisect, and this is the result:
```
There are only 'skip'ped commits left to test.
The first bad commit could be any of:
18a261999a822ce992ff260d96687ea1c76d4fd4
603bef2081c35ebde9bda959c8148fdfeb27d6ba
8711c20f46b49447fb26a40b83a7c017f985bd53
ca6d3994764f38bb55422148ab43036d7a5fc94e
451c2b7dda7fd7c38952737abc520a4709435bd6
We cannot bisect more!
```
I couldn't test the commits because even if the code compiled the filesystem didn't mount properly, but at least it should make the debug task a bit less daunting

**Comment by StrumentiResistenti** [comment]
Have you tried to reproduce the bug with an empty file having the same name? And what if you ```find``` the file in the archive directory of the repository? Do it exist twice or just once? 

The regression is probably related to the RDS layer I've introduced in 0.8.2. It's a sort of memory cache of datasets that prevents the same query from being resolved more than once. This happens especially while reading a directory, an operation that leads to N + 1 resolutions (one for the ```readdir``` and N for the ```getattr``` calls). The RDS gets populated on the ```readdir``` and later ```getattr``` can just access it to validate the existence of all the directory objects. Of course, the RDS layer suffers all the typical problems of a cache.

**Comment by conrad-heimbold** [comment]
Please solve this problem as soon as possible! It is extremely nerving! For me, it occured without changing the directory in nautilus (as far as I can remember). I just opened a video in one directory (via Nautilus) with Totem ; but when I closed Totem again and looked again to Nautilus windows, the file names had changed - from the normal ones to the internal.

**Comment by BrainDamage** [comment]
My apologies, I must have missed the reply notification, and I didn't bother checking manually for replies.

find lists the file only once:

```
$ find archive -iname "*Subaqueous Angels*"
archive/1/1___Ailsean - Subaqueous Angels of Chaos (FF I).mp3
```
same for every other duplicated file

> Have you tried to reproduce the bug with an empty file having the same name?
Yes, if done on the same repository, with the same tag path, after deleting the file, it produces a weird error:
```
$ touch "Ailsean - Subaqueous Angels of Chaos (FF I).mp3"
touch: setting times of 'Ailsean - Subaqueous Angels of Chaos (FF I).mp3': Bad address
```
The corruption seems to be random, when I create a new repository and copy over the same files, everything works for a while, then during normal daily operations I must do something that triggers the corruption, and from that point on I'll start seeing duplicates, the corrupted state has to be part of the written data since remounting doesn't reset it, and the glitched files are consistently the same, it also seems to get worse over time as now some files are even listed more than just 2 times in a row, even 4 for some

**Comment by StrumentiResistenti** [comment]
First of all I would ask you to query the DB to have a clearer picture of the mess that happened on filenames:

`select * from objects where objectname like "%Subaqueous Angels of Chaos%";`

Please adjust the query to retrieve only the duplicates of this file (or any other you see recurring more than once). I'm especially interested in inode and checksum. If you get duplicates on a single file, the mess is steadily inserted into the DB. Otherwise, we must assume that, given a repository, a set of circumstances duplicates some names in a reproducible way inside the RDS layer (Tagsistant cache). May be the order matters: you say that the same file is duplicated in one repo and not in another one made by tagging the same files in a different order.

Another request for you is to start Tagsistant with `--debug=cfFrs`, then to try to reproduce the duplicates and finally to post here the relevant rows from syslog. 

Thanks.

**Comment by BrainDamage** [comment]
I have good news: I managed to find a pattern in the broken files, it's partially my fault, but also tagsistant's reasoner's logic not handling query de-duplication from redundant tag relations: if tagA includes tagB and a file is tagged with both tagA and tagB it'll glitch out, I'd expect similar behavior with equivalence relationships

steps to reproduce:
```
$ cd /tmp
$ mkdir repo
mkdir: created directory 'repo'
$ tagsistant repo &>/dev/null
$ cd repo/tags
$ mkdir 1 2 3
mkdir: created directory '1'
mkdir: created directory '2'
mkdir: created directory '3'
$ cd ..
$ mkdir relations/1/includes/2
mkdir: created directory '2'
$ cd store/1/2/3/@
$ ls
$ touch broken
$ ls
1___broken  1___broken
```

**Comment by StrumentiResistenti** [comment]
Ah! This makes a lot of sense and helps us spotting the problem. 

Moreover, it can't be seen as your fault, since the user has the right to tag a file with the whole tagset and Tagsistant is responsible to report it just once, even if tag A includes B and tag C is equivalent to D and so on!

Thank you very much, I'll investigate this ASAP.

**Comment by StrumentiResistenti** [comment]
Here is the fix: a "distinct" clause was needed in a SQL query. Let me know if this works for you too.

**Comment by BrainDamage** [comment]
After some testing I can't reproduce the issue anymore, I'll close the ticket and eventually reopen if ever represents, thanks.
